core/common/src/main/java/alluxio/security/authentication/AuthenticatedChannelClientDriver.java

```diff
@@ -0,0 +1,195 @@
+import alluxio.grpc.GrpcChannelKey;
+import alluxio.grpc.SaslMessage;
+import alluxio.util.LogUtils;
+
+import com.google.common.util.concurrent.SettableFuture;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Responsible for driving authentication traffic from client-side.
+ *
+ * An authentication between client and server is managed by
+ * {@link AuthenticatedChannelClientDriver} and {@link AuthenticatedChannelServerDriver}
+ * respectively.
+ *
+ * These drivers are wrappers over gRPC {@link StreamObserver}s that manages the stream
+ * traffic destined for the other participant. They make sure messages are exchanged between client
+ * and server synchronously.
+ *
+ * Authentication is initiated by the client. Following the initiate call, depending on the scheme,
+ * one or more messages are exchanged to establish authenticated session between client and server.
+ *
+ * After the authentication is established, client and server streams are not closed in order to use
+ * them as long polling on authentication state changes.
+ *  -> Client closing the stream means that it doesn't want to be authenticated anymore.
+ *  -> Server closing the stream means the client is not authenticated at the server anymore.
+ *
+ */
+public class AuthenticatedChannelClientDriver implements StreamObserver<SaslMessage> {
+  private static final Logger LOG = LoggerFactory.getLogger(AuthenticatedChannelClientDriver.class);
+  /** Channel key. */
+  private GrpcChannelKey mChannelKey;
+  /** Server's sasl stream. */
+  private StreamObserver<SaslMessage> mRequestObserver;
+  /** Handshake handler for client. */
+  private SaslClientHandler mSaslClientHandler;
+  /** Whether channel is authenticated. */
+  private volatile boolean mChannelAuthenticated;
+  /** Used to wait during authentication handshake. */
+  private SettableFuture<Void> mChannelAuthenticatedFuture;
+
+  /**
+   * Creates client driver with given handshake handler.
+   *
+   * @param saslClientHandler sasl client handler
+   * @param channelKey channel key
+   */
+  public AuthenticatedChannelClientDriver(SaslClientHandler saslClientHandler,
+      GrpcChannelKey channelKey) {
+    mSaslClientHandler = saslClientHandler;
+    mChannelKey = channelKey;
+    mChannelAuthenticated = false;
+    mChannelAuthenticatedFuture = SettableFuture.create();
+  }
+
+  /**
+   * Sets the server's Sasl stream.
+   *
+   * @param requestObserver server Sasl stream
+   */
+  public void setServerObserver(StreamObserver<SaslMessage> requestObserver) {
+    mRequestObserver = requestObserver;
+  }
+
+  @Override
+  public void onNext(SaslMessage saslMessage) {
+    try {
+      LOG.debug("Received message for {}. Message: {}", mChannelKey.toStringShort(), saslMessage);
```

Comment:
This will invoke `toStringShort` even if it is not debug level, right? Instead, should we have a `if (debug())` gating this? (same with other `LOG.debug` statements here)

Reply:
We don't expect many calls to this. I'm in favor of keeping it simple.